docs/programmers_guide/witness_formal_spec.md

```diff
@@ -0,0 +1,196 @@
+# Block Witness Formal Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## The Stack
+
+Witnesses are executed on a stack machine that builds tries/calculates hashes.
+
+The stack consists of pairs `(node, hash)`.
+
+Each witness is a queue of instructions.
+In every execution cycle one instruction gets dequeued and a matching substitution rule gets applied to the stack.
+
+In the end, when there are no more instructions left, there should be only one item on the stack.  
+
+## Substitution rules
+
+Here is an example substitution rule. The current instruction is named `INSTR1`.
+
+```
+STACK(node1, hash1) STACK(node2, hash2) INSTR1(params) |=> 
+STACK(helper1(node1, node2), helper2(hash1, hash2))
+```
+
+This substitution rule replaces 2 stack items `(node1, hash1)` and `(node2, hash2)`
+with a single stack item `(helper1(node1, node2), helper2(hash1, hash2))`.
+
+Where `helper1` and `helper2` are pure functions defined in pseudocode (see the example below).
+
+```
+helper1 (value1, value2) {
+    return value1 + value2
+}
+
+helper2 (hash1, hash2) {
+    return KECCAK(hash1 + hash2)
+}
+```
+
+## Guards & Traps
+
+Each instruction and each helper function can have zero, one or multiple `guard` statements.
+Each `guard` statement looks like this:
+
+```
+guard <CONDITION> else {
+    TRAP
+}
+```
+
+That means that for this instruction to be valid the `<CONDITION>` in the guard statement must be true.
+
+If the condition is falsey, then the execution should stop (`TRAP` instruction). Implementation of the `TRAP` instruction is undefined, but it should stop the execution flow. (e.g. in Golang it might be a function returning an error or a panic).
```

Comment:
falsey => false. Is it possible to have have `TRAP` at all? Meaning that guard is just a condition

Reply:
but there are some unexpected cases, when there are no applicable substitution rules for instance or when there is more than one item left on stack after the execution.

I can try to get around this using `MUST` keyword though, but maybe it is more explicit to have some time of a trap condition.

Reply:
We can define in the semantics that if there are no applicable rules, then either we have 1 items left of the stack (which is the result), or the witness is not valid

Reply:
No, I don’t think it is how it works, because if we are searching for rules that means that we had an opcode already but it has invalid args most likely.